Q1. Explain the assumptions required to use ANOVA and provide examples of violations that could impact
the validity of the results.

The assumptions are 

1. Normality of sampling distribution.
i.e. the distribution of sample means are normally distributed.
2.Absence of outliers.Outlying scores need to be removed from the datset.
3.homogenity of variance.each one of the population should have equal variance.

If the assumption of normality is violated, or outliers are present, then the one-way ANOVA may not be the most powerful test available, and this could mean the difference between detecting a true difference among the population means or not.
As the variance used in ANOVA is the variance of the sample mean, it cannot be robust. One sizeable outlier may cause the sample mean to deviate from the true mean, and causes a substantial variation. Thus, the variance of the sample mean must not be considered robust.
If the homogeneity of variance assumption is failed, the ANOVA would yield a lower p-value, thus, leading us to falsely reject a true null hypothesis. If the independent assumption is violated, then also there is higher chance of getting an inaccurate P-value.


Q2. What are the three types of ANOVA, and in what situations would each be used?

There are 3 types of ANOVA:

1.One-way anova:one factor with atleast 2 levels. It is used when the levels ara independent.
2.Repeated measure anova: one factor with atleast 2 levels. It is used when the levels are dependent.
and
3. Factorial anova: two or mora factors each of which with atleast 2 levels,herethe levels can or cannot be dependent.

Q3. What is the partitioning of variance in ANOVA, and why is it important to understand this concept?

partition variance in ANOVA, that is, we could divide the total variance in our data into the various sources of that variation. Here, some of the variance results from variation among replicates within each group, and the rest comes from variation among the groups. Partitioning variance is an incredibly important concept within statistics. 

For scientists, it is a useful way of looking at the world.A range of scenarios use it to determine if there is any difference between the means of different groups.

Q4. How would you calculate the total sum of squares (SST), explained sum of squares (SSE), and residual
sum of squares (SSR) in a one-way ANOVA using Python?

We can get the total sum of square by adding sum of square between and sum of square within.
the formula is sum(Y**2)-T**2/N, where Y is value of every sample, T**2 is summation of square of each levels, and N is the number of total samples.
the sum of squares is a statistical technique used in regression analysis to determine the dispersion of data points. In a regression analysis, the goal is to determine how well a data series can be fitted to a function that might help to explain how the data series was generated. The sum of squares is used as a mathematical way to find the function that best fits (varies least) from the data.
The RSS measures the amount of error remaining between the regression function and the data set after the model has been run. A smaller RSS figure represents a regression function that is well-fit to the data.
The RSS, also known as the sum of squared residuals, essentially determines how well a regression model explains or represents the data in the model.
the formula for SSR is : RSS = sum over i = 1 to n (yi - f(xi))2
Where:
yi = the ith value of the variable to be predicted
f(xi) = predicted value of yi
n = upper limit of summation

Q5. In a two-way ANOVA, how would you calculate the main effects and interaction effects using Python?

In [62]:
#by the below example we can give the procedure how to calculate the main effect and interaction in a two way anova using python.
import pandas as pd
import numpy as np
import statsmodels.api as sm
from statsmodels.formula.api import ols
df = pd.DataFrame({'experience': np.repeat(['novice', 'experienced'], 15),
                   'program': np.tile(np.repeat(['A', 'B', 'C'], 5), 2),
                   'time': [6, 6, 6, 5, 6, 5, 5, 6, 4, 5,
                              6, 6, 7, 8, 7, 3, 4, 4, 4, 5,
                              4, 4, 4, 4, 4, 5, 6, 6, 7, 8]})
model = ols('time ~ C(experience) + C(program) + C(experience):C(program)', data=df).fit()
sm.stats.anova_lm(model, type=2)

df     sum_sq    mean_sq        F    PR(>F)
C(experience)              1.0   8.533333   8.533333  16.0000  0.000527
C(program)                 2.0  24.866667  12.433333  23.3125  0.000002
C(experience):C(program)   2.0   2.466667   1.233333   2.3125  0.120667
Residual                  24.0  12.800000   0.533333      NaN       NaN

Q6. Suppose you conducted a one-way ANOVA and obtained an F-statistic of 5.23 and a p-value of 0.02.
What can you conclude about the differences between the groups, and how would you interpret these
results?

F-statistic = 5.23
p_value = .02
as f-statistic value is greaater than p-value, i.e. we will be rejecting the null hypothesis i.e. there will be at leaste one variance which will be unequal

Q7. In a repeated measures ANOVA, how would you handle missing data, and what are the potential
consequences of using different methods to handle missing data?

The three main options to handle missing data are:
Missing Completely at Random (MCAR) - i.e. there are no possible predictors of missingness. In this case, we could validly drop those individuals with missing data and work with the complete observations only (this is called a complete case analysis). However, this is generally an unrealistic assumption in my experience. If we do a complete case analysis and the MCAR assumption does not hold, then any conclusions we want to draw from our results will only be valid for that specific set of people who have complete data. It is invalid to make inference about the whole population (or even whole sample) if MCAR does not hold, when using complete case analysis.

Missing at Random (MAR) - in this case, we know why some individuals have missing data and others do not (there are some variables which predict missingness) but within strata of those variables the data is truly randomly missing. In this case, we can use multiple imputation or other methods such as inverse probability weighting for non-missingness (IPW). Under MAR, these methods allow us to make inference about the entire sample, and therefore about the population from which your samples arose. we are right that you have a lot of data to impute, which may cause a problem.

Not Missing at Random (NMAR) - there is a definite pattern to which data is missing and there are no variables which can be used to create strata within which the data is missing at random.

Q9. A researcher wants to compare the mean weight loss of three diets: A, B, and C. They collect data from
50 participants who were randomly assigned to one of the diets. Conduct a one-way ANOVA using Python
to determine if there are any significant differences between the mean weight loss of the three diets.
Report the F-statistic and p-value, and interpret the results.

In [63]:
from scipy.stats import f_oneway
A = [8,7,4,5,9,8,6,3,2,5,1,4,8,7,5,2]
B = [3, 2, 4, 9, 8,4,7,8,5,6,9,8,5,4,1,2]
C = [9, 8, 9, 3, 9,4,5,8,6,9,3,2,5,4,7,1,2,4]
f_oneway(A, B, C)

F_onewayResult(statistic=0.024604366060866762, pvalue=0.9757084126578006)

In [64]:
#As f statistic is not greater than p value, i.e. we will fail to reject the null hypotheis, i.e. there is same  weight in each group.

Q10. A company wants to know if there are any significant differences in the average time it takes to
complete a task using three different software programs: Program A, Program B, and Program C. They
randomly assign 30 employees to one of the programs and record the time it takes each employee to
complete the task. Conduct a two-way ANOVA using Python to determine if there are any main effects or
interaction effects between the software programs and employee experience level (novice vs.
experienced). Report the F-statistics and p-values, and interpret the results.

In [1]:

import pandas as pd
import numpy as np
import statsmodels.api as sm
from statsmodels.formula.api import ols
df = pd.DataFrame({'experience': np.repeat(['novice', 'experienced'], 15),
                   'program': np.tile(np.repeat(['A', 'B', 'C'], 5), 2),
                   'time': [6, 6, 6, 5, 6, 5, 5, 6, 4, 5,
                              6, 6, 7, 8, 7, 3, 4, 4, 4, 5,
                              4, 4, 4, 4, 4, 5, 6, 6, 7, 8]})
model = ols('time ~ C(experience) + C(program) + C(experience):C(program)', data=df).fit()
sm.stats.anova_lm(model, type=2)

df     sum_sq    mean_sq        F    PR(>F)
C(experience)              1.0   8.533333   8.533333  16.0000  0.000527
C(program)                 2.0  24.866667  12.433333  23.3125  0.000002
C(experience):C(program)   2.0   2.466667   1.233333   2.3125  0.120667
Residual                  24.0  12.800000   0.533333      NaN       NaN

experienced: p-value = .000527
program: p-value = .0000002
experienced*program: p-value = .120667

Since the p-values for experience and program are both less than .05, this means that both factors have a statistically significant effect on time.

And since the p-value for the interaction effect (.120667) is not less than .05, this tells us that there is no significant interaction effect between program and experience.

Q11. An educational researcher is interested in whether a new teaching method improves student test
scores. They randomly assign 100 students to either the control group (traditional teaching method) or the
experimental group (new teaching method) and administer a test at the end of the semester. Conduct a
two-sample t-test using Python to determine if there are any significant differences in test scores
between the two groups. If the results are significant, follow up with a post-hoc test to determine which
group(s) differ significantly from each other.

In [66]:
import numpy as np
import scipy.stats as stats
control_group = np.array([77,45,78,96,75,48,63,40,15,75,85,95,88,69,45,87,36,52,85,98,75,84,63,56,54,84,32,89,69,47,25,48,36,87,95,84,62,59,58,47,69,26,45,85,96,87,36,54,85,78])
experimental_group = np.array([78,96,75,48,63,40,15,75,85,95,88,69,45,87,36,78,58,75,48,99,47,85,47,96,84,87,85,83,96,95,98,97,94,92,95,96,36,35,35,34,32,37,38,39,41,44,75,48,46,77])
print(np.var(control_group), np.var(experimental_group))
stats.ttest_ind(a=control_group, b=experimental_group, equal_var=True)

475.06440000000003 593.3364


Ttest_indResult(statistic=-0.3426501453665429, pvalue=0.7325955227430436)

The two hypotheses for this particular two sample t-test are as follows:
H0: µ1 = µ2
HA: µ1 ≠µ2
Because the p-value of our test (0.7325955227430436) is greater than alpha = 0.05, we fail to reject the null hypothesis of the test.

Q12. A researcher wants to know if there are any significant differences in the average daily sales of three
retail stores: Store A, Store B, and Store C. They randomly select 30 days and record the sales for each store
on those days. Conduct a repeated measures ANOVA using Python to determine if there are any

significant differences in sales between the three stores. If the results are significant, follow up with a post-
hoc test to determine which store(s) differ significantly from each other.

In [67]:
import numpy as np
import pandas as pd
df = pd.DataFrame({'store': np.repeat(['A', 'B','C'], 10),
                   'day': np.tile([1,2,3,4,5,6,7,8,9,10], 3),
                   'sales': [30, 28, 16, 34,
                                14, 18, 10, 22,
                                24, 20, 18, 30,
                                38, 34, 20, 44, 
                                26, 28, 14, 30,
                                45, 85, 96, 34,
                                78 ,85, 96, 78,45,96]})

from statsmodels.stats.anova import AnovaRM
print(AnovaRM(data=df, depvar='sales', subject='store', within=['day']).fit())


              Anova
    F Value Num DF  Den DF Pr > F
---------------------------------
day  0.8030 9.0000 18.0000 0.6191



In this example, the F test-statistic is 0.8030 and the corresponding p-value is 0.6191.
Since this p-value is less than 0.05, we reject the null hypothesis and conclude that there is a statistically significant difference between three stores..
